pkg/tracing/otlp_sdk.go

```diff
@@ -0,0 +1,112 @@
+package tracing
+
+import (
+	"context"
+	"os"
+	"time"
+
+	"go.opentelemetry.io/contrib/exporters/autoexport"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+
+	"github.com/go-logr/logr"
+	errors "github.com/zgalor/weberr"
+)
+
+// ConfigureOpenTelemetryTracer configures the global OpenTelemetry trace
+// provider and propagator for baggages and trace context.
+//
+// The function uses the following environment variables for the tracer
+// configuration:
+//   - `OTEL_TRACES_EXPORTER`, either `otlp` to send traces to an OTLP endpoint or `console`.
+//   - `OTEL_EXPORTER_OTLP_TRACES_PROTOCOL`, either `grpc` or `http`.
+//   - `OTEL_EXPORTER_OTLP_TRACES_ENDPOINT`, endpoint where to send the OTLP
+//     traces (e.g. `https://localhost:4318/v1/traces`).
+//
+// See
+// https://pkg.go.dev/go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp
+// for the list of all supported variables.
+//
+// An error is returned if an environment value is set to an unhandled value.
+//
+// If no environment variable are set, a no-op tracer is setup.
+func ConfigureOpenTelemetryTracer(ctx context.Context, logger logr.Logger, resourceAttrs ...attribute.KeyValue) (func(context.Context) error, error) {
+	logger.Info("initializing OpenTelemetry tracer...")
+
+	exp, err := autoexport.NewSpanExporter(ctx, autoexport.WithFallbackSpanExporter(newNoopFactory))
+	if err != nil {
+		return nil, errors.Errorf("failed to create OTEL exporter: %s", err)
+	}
+
+	var isNoop bool
+	if _, isNoop = exp.(*noopSpanExporter); !isNoop || autoexport.IsNoneSpanExporter(exp) {
+		isNoop = true
+	}
+	logger.Info("initializing OpenTelemetry tracer:", "isNoop", isNoop)
+
+	opts := []resource.Option{resource.WithHost()}
+	if len(resourceAttrs) > 0 {
+		opts = append(opts, resource.WithAttributes(resourceAttrs...))
+	}
+	resources, err := resource.New(ctx, opts...)
+	if err != nil {
+		return nil, errors.Errorf("failed to initialize trace resources: %s", err)
+	}
+
+	tp := tracesdk.NewTracerProvider(
+		tracesdk.WithBatcher(exp),
+		tracesdk.WithResource(resources),
+	)
+	otel.SetTracerProvider(tp)
+
+	shutdown := func(ctx context.Context) error {
+		ctx, cancel := context.WithTimeout(ctx, 5*time.Second)
+		defer cancel()
+		return tp.Shutdown(ctx)
+	}
+
+	propagator := propagation.NewCompositeTextMapPropagator(propagation.Baggage{}, propagation.TraceContext{})
+	otel.SetTextMapPropagator(propagator)
+
+	otel.SetErrorHandler(otelErrorHandlerFunc(func(err error) {
+		logger.Error(err, "OpenTelemetry.ErrorHandler: %v", err.Error())
+	}))
+
+	return shutdown, nil
+}
+
+// TracingEnabled returns true if the environment variable OTEL_TRACES_EXPORTER
+// to configure the OpenTelemetry Exporter is defined.
+func TracingEnabled() bool {
+	_, ok := os.LookupEnv("OTEL_TRACES_EXPORTER")
```

Comment:
I am wondering if this package should really be the responsible of deciding in what specific way the tracing is enabled. Some services might want to enable it through the standardized environment variable and some other services might not want to. For example, they might decide it is a configuration file attribute they specify instead, because they have the configuration centralized.

In one of the services where we configured tracing I left a comment in a MR with details about this insight. I am not putting the link here as it points to internal information but you can reach back to me and/or @simonpasquier with details about that conversation.

Reply:
The same comment applies to other environment variables

Reply:
This is a fair point. I will be fine removing this part off the package and let the user decide how they want to enable tracing and then ultilize this package to configure the tracing.
WDYT? @simonpasquier

Reply:
Simon and I discussed about this and we agreed to remove this piece of code. Users can check for tracing enablement using whatever method that suits them. Then, they can use this library to configure the tracing.